includes/sanitizers/class-amp-href-sanitizer.php

```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * Class AMP_Href_Sanitizer.
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Href_Sanitizer.
+ *
+ * Crawls all href attributes to ensure their URLs are valid.
+ */
+final class AMP_Href_Sanitizer extends AMP_Base_Sanitizer {
+
+	/**
+	 * Xpath query selector to target all elements with a href attribute.
+	 *
+	 * @var string
+	 */
+	const XPATH_SELECTOR = '//*[@href]';
+
+	/**
+	 * Array of attribute names that should be removed for invalid hrefs.
+	 *
+	 * @var array<string>
+	 */
+	private static $attributes_tied_to_href = [
+		'target',
+		'download',
+		'rel',
+		'rev',
+		'hreflang',
+		'type',
+	];
+
+	/**
+	 * Sanitize the HTML contained in the DOMDocument received by the
+	 * constructor
+	 *
+	 * @throws LogicException If the xpath query produced invalid results.
+	 */
+	public function sanitize() {
+		$xpath = new DOMXPath( $this->dom );
+
+		foreach ( $xpath->query( self::XPATH_SELECTOR ) as $node ) {
+			// This should not happen, something is wrong with our query.
+			if ( ! $node instanceof DOMElement
+				|| ! $node->hasAttribute( 'href' ) ) {
+				throw new LogicException(
```

Comment:
Do we use asserts and logic exceptions for defensive development to make the code more resilient (and document it at the same time)?

Reply:
As discussed in our chat today, the XPath for `*[@href]` will never return anything but elements containing `href` attributes, so this logic is unnecessary/redundant.

Reply:
https://github.com/ampproject/amp-wp/pull/3117/commits/15fa5154e1ab5e9199d997e5a1819cc7342e299a